Client/Frontend/Browser/FrecencyQuery.swift

```diff
@@ -0,0 +1,38 @@
+// Copyright 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Shared
+import Storage
+import Data
+
+class FrecencyQuery {
+    private static let queue = DispatchQueue(label: "frecency-query-queue")
+    private static var cancellable: DispatchWorkItem?
+    
+    public static func sitesByFrecency(containing query: String? = nil,
+                                       completion: @escaping (Set<Site>) -> Void) {
+        cancellable?.cancel()
+        cancellable = nil
+        
+        // CoreData is fast, can be fetched on main thread. This also prevents threading problems.
+        let historySites = History.byFrecency(query: query)
+            .map { Site(url: $0.url ?? "", title: $0.title ?? "") }
+        cancellable = DispatchWorkItem {
+            // brave-core fetch can be slow over 200ms per call,
+            // a cancellable serial queue is used for it.
+            let bookmarkSites = Bookmarkv2.byFrequency(query: query)
+                .map { Site(url: $0.url ?? "", title: $0.title ?? "", bookmarked: true) }
+            
+            let result = Set<Site>(historySites + bookmarkSites)
+            
+            completion(result)
```

Comment:
Why not dispatch this to main in this function rather than in the call site?

Reply:
I'll change it,
I did not dispatch on main here because we do additional processing after completion:
```
for site in result {
                    if let completion = self.completionForURL(site.url) {
```

My goal was to move as much code as possible on background thread to reduce UI lag, this code is called on each key stroke

Reply:
Gotcha, yeah I saw there was further processing, just wanted to make sure it was intentional since it breaks usual convention. If we're keeping it maybe just add documentation mentioning it calls completion on a background queue